### ISSUE TITLE
Add CI tests for rusty-hermit running on uhyve.

### ISSUE THREAD

**Issue by jschwe** [hermitcore/rusty-hermit#2]
Currently it seems that all tests only use rusty-hermit on Qemu. Travis CI has nested virtualization enabled by default (seems to be a recent change). This enables us to test rusty-hermit using uhyve.

I'm currently working on a travis pipeline and I'll be opening a pull request for this. 
Meanwhile I'd like to link the following two uhyve issues here, which are important for CI.

- https://github.com/hermitcore/uhyve/issues/6
- https://github.com/hermitcore/uhyve/issues/7

**Comment by jschwe** [comment]
I'm currently working on two Travis CI pipelines. 
The first one builds everything as described in the Readme, i.e. 

- Build rusty-hermit demo with rust nightly from rustup and run with uhyve from cargo

The second pipeline compiles everything from source so that regressions due to changes in uhyve or in the rust repo can be quickly identified.
- The rust compiler fork from hermitcore is used
- uhyve and rusty-hermit are built from source with the the newly built compiler
- run rusty-hermit demo on uhyve

@stlankes What do you think about this? And which method of pinning the dependencies (rust and uhyve) do you prefer? I've thought of the following options:

- Add another repo that includes rusty-hermit, hermitcore/rust and uhyve as submodules. (This is the solution that I'm currently working with)
- Add rust and uhyve as submodules to this repo
- Make a config file in this repo
- Hardcode the branches that should be checked out for rust and uhyve in the travis pipeline ( This is a bit too unflexible in my opinion)

**Comment by stlankes** [comment]
@jschwe I removed Travis CI support because we have our own [pipeline](https://git.rwth-aachen.de/acs/public/hermitcore/rusty-hermit/pipelines). Using uhyve is not so easy in CI environment. As fas as I know, Travis CI doesn't support nested virtualization. In our environment, we are able to support nested virtualization. It is currently deactivated because we are reconfiguring our setup. @stv0g is our CI expert.

We provide also the compiler as [docker image](https://git.rwth-aachen.de/acs/public/hermitcore/hermitrust). But this image based on the original nightly compiler and doesn't run any tests.

**Comment by jschwe** [comment]
Of the major CI providers to my knowledge only Travis CI provides nested virtualization by default. This seems to be somewhat recent too (late 2019 or early 2020 from what I can tell). 
Based on your comment I'm assuming you are working on configuring your self hosted gitlab runner to provide nested virtualization, is this correct? Using that would of course be preferred, I just chose Travis because it works by default.

Basically this issue also isn't about which CI provider should be used, but I think it would be good if there where a CI test that tests rusty-hermit running on uhyve, both on current nightly rust builds as well as the current compiler and uhyve version in the repository. This requires specifying the dependency in some form, e.g. via submodules or a config file.

**Comment by stlankes** [comment]
I agree. I will talk @stv0g if we are enable tests with uhyve on our CI.

**Comment by stlankes** [comment]
@jschwe Do we solve this issue?

**Comment by jschwe** [comment]
Ah yes, thanks for reminding me. CI tests with uhyve were added so I'm closing the issue.
